app/scripts/lib/rpc-method-middleware/handlers/provider-authorize.js

```diff
@@ -0,0 +1,200 @@
+//     "caip154-mandatory": "true"
+//   }
+// }
+
+const providerAuthorize = {
+  methodNames: [MESSAGE_TYPE.PROVIDER_AUTHORIZE],
+  implementation: providerAuthorizeHandler,
+  hookNames: {
+    getAccounts: true,
+  },
+};
+export default providerAuthorize;
+
+async function providerAuthorizeHandler(_req, res, _next, end, _hooks) {
+  const { requiredScopes, optionalScopes, sessionProperties, ...restParams } =
+    _req.params;
+
+  if (Object.keys(restParams).length !== 0) {
+    return end(
+      new EthereumRpcError(
+        5301,
+        'Session Properties can only be optional and global',
+      ),
+    );
+  }
+
+  const sessionId = '0xdeadbeef';
+
+  const validRequiredScopes = {};
+  for (const [scopeString, scopeObject] of Object.entries(requiredScopes)) {
+    if (isValidScope(scopeString, scopeObject)) {
+      validRequiredScopes[scopeString] = {
+        accounts: [],
+        ...scopeObject,
+      };
+    }
+  }
+  if (requiredScopes && Object.keys(validRequiredScopes).length === 0) {
+    // What error code and message here?
+    throw new Error(
+      '`requiredScopes` object MUST contain 1 more `scopeObjects`, if present',
+    );
+  }
+
+  const validOptionalScopes = {};
+  for (const [scopeString, scopeObject] of Object.entries(optionalScopes)) {
+    if (isValidScope(scopeString, scopeObject)) {
+      validOptionalScopes[scopeString] = {
+        accounts: [],
+        ...scopeObject,
+      };
+    }
+  }
+  if (optionalScopes && Object.keys(validOptionalScopes).length === 0) {
+    // What error code and message here?
+    throw new Error(
+      '`optionalScopes` object MUST contain 1 more `scopeObjects`, if present',
+    );
+  }
+
+  const randomSessionProperties = {}; // session properties do not have to be honored by the wallet
+  for (const [key, value] of Object.entries(sessionProperties)) {
+    if (Math.random() > 0.5) {
+      randomSessionProperties[key] = value;
+    }
+  }
+  if (sessionProperties && Object.keys(sessionProperties).length === 0) {
+    return end(
+      new EthereumRpcError(5300, 'Invalid Session Properties requested'),
+    );
+  }
+
+  const validScopes = {
+    // what happens if these keys collide?
+    ...validRequiredScopes,
+    ...validOptionalScopes,
+  };
+
+  // TODO:
+  // Unless the dapp is known and trusted, give generic error messages for
+  // - the user denies consent for exposing accounts that match the requested and approved chains,
+  // - the user denies consent for requested methods,
+  // - the user denies all requested or any required scope objects,
+  // - the wallet cannot support all requested or any required scope objects,
+  // - the requested chains are not supported by the wallet, or
+  // - the requested methods are not supported by the wallet
+  // return
+  //     "code": 0,
+  //     "message": "Unknown error"
+
+  if (Object.keys(validScopes).length === 0) {
+    return end(new EthereumRpcError(5000, 'Unknown error with request'));
```

Comment:
I say we just give them the least amount of info, the 'ol generic internal server error.

```suggestion
    return end(new EthereumRpcError(-32603, 'Internal error'));
```

Reply:
this comes from spec

https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-25.md#:~:text=message%20%3D%20%22Unknown%20error%20with%20request%22